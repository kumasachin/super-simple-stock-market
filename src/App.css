.App {
  text-align: center;
  font-family: Arial, Helvetica, sans-serif;
}

.tableFormat {
  border-collapse: collapse;
  width: 90%;
  margin:0 auto;
}

.tableFormat td, .tableFormat th {
  border: 1px solid #ddd;
  padding: 8px;
}

.tableFormat tr:nth-child(even){background-color: #f2f2f2;}

.tableFormat tr:hover {background-color: #ddd;}

.tableFormat th {
  padding-top: 12px;
  padding-bottom: 12px;
  text-align: left;
  background-color: #4CAF50;
  color: white;
}