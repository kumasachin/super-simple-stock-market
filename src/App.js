import React, { useState } from "react";
import {
  Stocks,
  Trades
} from "./components/";
import stockData from "./data/stock";
import './App.css';
import {
  nthRoot
} from "./utils/common";

const App = () => {
  const [stocksData, setStocksData] = useState(stockData);
  const [tradesData, setTradesData] = useState([]);
  const [geomatricMean, setGeomatricMean] = useState("");
  const setStockInStore = (stock, index) => {
    const stockLocalData = [
      ...stocksData
    ];
    stockLocalData[index] = stock;
    setStocksData(stockLocalData);
  };
  const getGeomatricMean = () => {
    const filledElementArray = stocksData.map(stock => stock.price).filter(Boolean);
    if(filledElementArray.length > 0) {
      const allPriceMultplied = filledElementArray.reduce((prevPrice, nextPrice) => {
        if (prevPrice) {
          return prevPrice*nextPrice;
        }
        return nextPrice;    
      }, 0);
      const geomatricMean = nthRoot(allPriceMultplied, filledElementArray.length);
      setGeomatricMean(geomatricMean);
    } else {
      setGeomatricMean("");
    }
  };

  const handleTrade = (trade) => {
    setTradesData([
      ...tradesData, 
      trade
    ])
  };

  return (
    <div className="App">
      <header>
        <h1 data-testid="application-heading">Super simple stock market</h1>
      </header>
      <Stocks
        trades={tradesData} 
        stocks={stocksData} 
        setStockInStore={setStockInStore} 
        handleTrade={handleTrade} 
      />
      <br/>
      <h3 data-testid="geo-matric-mean-container">
        {
        geomatricMean ? (
          <span data-testid="geo-matric-mean">
            Current Geomatric Mean = {
              geomatricMean.toLocaleString(undefined, {
                minimumFractionDigits: 2,
                maximumFractionDigits: 2
              })
            }
          </span>): null 
        }
      </h3>
      <button data-testid="btn-get-geomatric-mean" onClick={getGeomatricMean}>Calculate Geometric Mean</button>
      <br/>
      {
        tradesData.length > 0 && <Trades 
          trades={tradesData} 
        />
      }
    </div>
  );
}

export default App;
