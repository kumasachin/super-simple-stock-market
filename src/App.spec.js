import { render, screen, fireEvent, waitFor, cleanup } from '@testing-library/react';
import App from './App';

describe("App", () => {
  afterEach(cleanup);
  beforeEach(()=> render(<App />));
  it('should take a snapshot', () => {
    const { asFragment } = render(<App />);
    expect(asFragment(<App />)).toMatchSnapshot()
  });

  it('should calculate no geomatricmean if all input are blank', async () => {
    const btnGetGeomatricMean = screen.getByTestId("btn-get-geomatric-mean");
    fireEvent.click(btnGetGeomatricMean);
    expect(screen.getByTestId("geo-matric-mean-container")).toHaveTextContent("");
  });

  it('should calculate geomatricmean', async () => {
    const pricefield = screen.getByTestId("stock-price-0");
    fireEvent.blur(pricefield, { target: { value: "123" }});
    const btnGetGeomatricMean = screen.getByTestId("btn-get-geomatric-mean");
    fireEvent.click(btnGetGeomatricMean);
    expect(screen.getByTestId("geo-matric-mean")).toHaveTextContent("Current Geomatric Mean = 123.00");
  });

  it('should calculate geomatricmean of more that 2 inputs', async () => {
    fireEvent.blur(screen.getByTestId("stock-price-0"), { target: { value: "223" }});
    fireEvent.blur(screen.getByTestId("stock-price-1"), { target: { value: "223" }});
    const btnGetGeomatricMean = screen.getByTestId("btn-get-geomatric-mean");
    fireEvent.click(btnGetGeomatricMean);
    expect(screen.getByTestId("geo-matric-mean")).toHaveTextContent("Current Geomatric Mean = 223.00");
  });

  it('should not default activate the buy and sell button', async () => {
    const buybtn = screen.getByTestId("stock-buy-0");
    const sellbtn = screen.getByTestId("stock-sell-0");
    expect(buybtn).not.toBeEnabled();
    expect(sellbtn).not.toBeEnabled();
  });

  it('should activate the buy and sell button', async () => {
    const buybtn = screen.getByTestId("stock-buy-0");
    const sellbtn = screen.getByTestId("stock-sell-0");
    expect(buybtn).not.toBeEnabled();
    expect(sellbtn).not.toBeEnabled();
  });

  it('should activate buy/ sell button and render trade rows for buy/sell', async () => {
    const buybtn = screen.getByTestId("stock-buy-0");
    const sellbtn = screen.getByTestId("stock-sell-0");
    expect(buybtn).not.toBeEnabled();
    expect(sellbtn).not.toBeEnabled();

    fireEvent.blur(screen.getByTestId("stock-price-0"), { target: { value: "223" }});
    fireEvent.blur(screen.getByTestId("stock-quantity-0"), { target: { value: "2" }});
    fireEvent.change(screen.getByTestId("stock-quantity-0"), { target: { value: "2" }});
    expect(buybtn).toBeEnabled();
    expect(sellbtn).toBeEnabled();
    expect(screen.getByTestId("geo-matric-mean-container")).toHaveTextContent("");
    
     waitFor(() => fireEvent.click(buybtn));
    expect(screen.getByTestId("trade-dateandtime-0")).toBeTruthy();
    expect(screen.getByTestId("trade-symbol-0")).toHaveTextContent("TEA");
    expect(screen.getByTestId("trade-quanity-0")).toHaveTextContent("2");
    expect(screen.getByTestId("trade-action-0")).toHaveTextContent('buy');
    expect(screen.getByTestId("trade-price-0")).toHaveTextContent('223.00');
    expect(screen.getByTestId("trade-totalprice-0")).toHaveTextContent('446.00'); 

    const buybtn02 = screen.getByTestId("stock-buy-2");
    const sellbtn02 = screen.getByTestId("stock-sell-2");
    expect(buybtn02).not.toBeEnabled();
    expect(sellbtn02).not.toBeEnabled();

    fireEvent.blur(screen.getByTestId("stock-price-2"), { target: { value: "12223" }});
    fireEvent.blur(screen.getByTestId("stock-quantity-2"), { target: { value: "12" }});
    fireEvent.change(screen.getByTestId("stock-quantity-2"), { target: { value: "12" }});
    expect(buybtn02).toBeEnabled();
    expect(sellbtn02).toBeEnabled();
    expect(screen.getByTestId("geo-matric-mean-container")).toHaveTextContent("");
    waitFor(() => fireEvent.click(sellbtn02));

    expect(screen.getByTestId("trade-dateandtime-1")).toBeTruthy();
    expect(screen.getByTestId("trade-symbol-1")).toHaveTextContent("ALE");
    expect(screen.getByTestId("trade-quanity-1")).toHaveTextContent("2");
    expect(screen.getByTestId("trade-action-1")).toHaveTextContent("sell");
    expect(screen.getByTestId("trade-price-1")).toHaveTextContent("223.00");
    expect(screen.getByTestId("trade-totalprice-1")).toHaveTextContent("146,676.00");
  });
});
