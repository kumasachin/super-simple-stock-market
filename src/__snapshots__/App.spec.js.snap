// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <header>
      <h1
        data-testid="application-heading"
      >
        Super simple stock market
      </h1>
    </header>
    <h2
      data-testid="stock-table-title"
    >
      Stocks
    </h2>
    <table
      class="tableFormat"
      data-testid="stock-table"
    >
      <thead
        data-testid="stock-thead"
      >
        <tr
          data-testid="stock-tr"
        >
          <th>
            Stock Symbol
          </th>
          <th>
            Type
          </th>
          <th>
            Last Dividend
          </th>
          <th>
            Fixed Dividend
          </th>
          <th>
            Par Value
          </th>
          <th>
            Price
          </th>
          <th
            width="15%"
          >
            Dividend Yield
          </th>
          <th
            width="15%"
          >
            P/E Ratio
          </th>
          <th
            width="15%"
          >
            Volume Weighted Stock Price
          </th>
          <th>
            Quantity
          </th>
          <th>
            Action
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          data-testid="stock-row-0"
        >
          <td
            data-testid="stock-symbol-0"
          >
            TEA
          </td>
          <td
            data-testid="stock-type-0"
          >
            Common
          </td>
          <td
            data-testid="stock-lastdivident-0"
          >
            -
          </td>
          <td
            data-testid="stock-fixeddividend-0"
          >
            -
          </td>
          <td
            data-testid="stock-parvalue-0"
          >
            100
          </td>
          <td>
            <input
              data-testid="stock-price-0"
              min="1"
              name="price"
              type="number"
            />
          </td>
          <td
            data-testid="stock-dividentyield-0"
          />
          <td
            data-testid="stock-peratio-0"
          />
          <td
            data-testid="stock-vwsprice-0"
          />
          <td>
            <input
              data-testid="stock-quantity-0"
              min="1"
              name="quantity"
              type="number"
              value=""
            />
          </td>
          <td
            data-testid="stock-buysell-0"
          >
            <button
              data-testid="stock-buy-0"
              disabled=""
              name="buy"
            >
              Buy
            </button>
            <br />
            <br />
            <button
              data-testid="stock-sell-0"
              disabled=""
              name="sell"
            >
              Sell
            </button>
          </td>
        </tr>
        <tr
          data-testid="stock-row-1"
        >
          <td
            data-testid="stock-symbol-1"
          >
            POP
          </td>
          <td
            data-testid="stock-type-1"
          >
            Common
          </td>
          <td
            data-testid="stock-lastdivident-1"
          >
            8
          </td>
          <td
            data-testid="stock-fixeddividend-1"
          >
            -
          </td>
          <td
            data-testid="stock-parvalue-1"
          >
            100
          </td>
          <td>
            <input
              data-testid="stock-price-1"
              min="1"
              name="price"
              type="number"
            />
          </td>
          <td
            data-testid="stock-dividentyield-1"
          />
          <td
            data-testid="stock-peratio-1"
          />
          <td
            data-testid="stock-vwsprice-1"
          />
          <td>
            <input
              data-testid="stock-quantity-1"
              min="1"
              name="quantity"
              type="number"
              value=""
            />
          </td>
          <td
            data-testid="stock-buysell-1"
          >
            <button
              data-testid="stock-buy-1"
              disabled=""
              name="buy"
            >
              Buy
            </button>
            <br />
            <br />
            <button
              data-testid="stock-sell-1"
              disabled=""
              name="sell"
            >
              Sell
            </button>
          </td>
        </tr>
        <tr
          data-testid="stock-row-2"
        >
          <td
            data-testid="stock-symbol-2"
          >
            ALE
          </td>
          <td
            data-testid="stock-type-2"
          >
            Common
          </td>
          <td
            data-testid="stock-lastdivident-2"
          >
            23
          </td>
          <td
            data-testid="stock-fixeddividend-2"
          >
            -
          </td>
          <td
            data-testid="stock-parvalue-2"
          >
            60
          </td>
          <td>
            <input
              data-testid="stock-price-2"
              min="1"
              name="price"
              type="number"
            />
          </td>
          <td
            data-testid="stock-dividentyield-2"
          />
          <td
            data-testid="stock-peratio-2"
          />
          <td
            data-testid="stock-vwsprice-2"
          />
          <td>
            <input
              data-testid="stock-quantity-2"
              min="1"
              name="quantity"
              type="number"
              value=""
            />
          </td>
          <td
            data-testid="stock-buysell-2"
          >
            <button
              data-testid="stock-buy-2"
              disabled=""
              name="buy"
            >
              Buy
            </button>
            <br />
            <br />
            <button
              data-testid="stock-sell-2"
              disabled=""
              name="sell"
            >
              Sell
            </button>
          </td>
        </tr>
        <tr
          data-testid="stock-row-3"
        >
          <td
            data-testid="stock-symbol-3"
          >
            GIN
          </td>
          <td
            data-testid="stock-type-3"
          >
            Preferred
          </td>
          <td
            data-testid="stock-lastdivident-3"
          >
            8
          </td>
          <td
            data-testid="stock-fixeddividend-3"
          >
            2%
          </td>
          <td
            data-testid="stock-parvalue-3"
          >
            100
          </td>
          <td>
            <input
              data-testid="stock-price-3"
              min="1"
              name="price"
              type="number"
            />
          </td>
          <td
            data-testid="stock-dividentyield-3"
          />
          <td
            data-testid="stock-peratio-3"
          />
          <td
            data-testid="stock-vwsprice-3"
          />
          <td>
            <input
              data-testid="stock-quantity-3"
              min="1"
              name="quantity"
              type="number"
              value=""
            />
          </td>
          <td
            data-testid="stock-buysell-3"
          >
            <button
              data-testid="stock-buy-3"
              disabled=""
              name="buy"
            >
              Buy
            </button>
            <br />
            <br />
            <button
              data-testid="stock-sell-3"
              disabled=""
              name="sell"
            >
              Sell
            </button>
          </td>
        </tr>
        <tr
          data-testid="stock-row-4"
        >
          <td
            data-testid="stock-symbol-4"
          >
            GIN
          </td>
          <td
            data-testid="stock-type-4"
          >
            Common
          </td>
          <td
            data-testid="stock-lastdivident-4"
          >
            13
          </td>
          <td
            data-testid="stock-fixeddividend-4"
          >
            -
          </td>
          <td
            data-testid="stock-parvalue-4"
          >
            250
          </td>
          <td>
            <input
              data-testid="stock-price-4"
              min="1"
              name="price"
              type="number"
            />
          </td>
          <td
            data-testid="stock-dividentyield-4"
          />
          <td
            data-testid="stock-peratio-4"
          />
          <td
            data-testid="stock-vwsprice-4"
          />
          <td>
            <input
              data-testid="stock-quantity-4"
              min="1"
              name="quantity"
              type="number"
              value=""
            />
          </td>
          <td
            data-testid="stock-buysell-4"
          >
            <button
              data-testid="stock-buy-4"
              disabled=""
              name="buy"
            >
              Buy
            </button>
            <br />
            <br />
            <button
              data-testid="stock-sell-4"
              disabled=""
              name="sell"
            >
              Sell
            </button>
          </td>
        </tr>
      </tbody>
    </table>
    <br />
    <h3
      data-testid="geo-matric-mean-container"
    />
    <button
      data-testid="btn-get-geomatric-mean"
    >
      Calculate Geometric Mean
    </button>
    <br />
  </div>
</DocumentFragment>
`;
